### DISCUSSION TITLE
Botorch Tensor Dimension Error

### DISCUSSION THREAD

**Discussion by kks2107** [meta-pytorch/botorch#1514] [Q&A] [answered]
We have a problem with Botorch. Using Botorch for the first time and now we get the error "BotorchTensorDimensionError: pareto_Y must have the same number of objectives as ref_point. Got 4, expected 2.".
But the size should not be the same.

**Comment by saitcakmak** [comment, marked as answer]
Hi @kks2107. Can you share the stack trace? Where do you get the error and what are the inputs to the function that produces the error?

Based on the error message alone, I assume it is raised in `hv.compute(...)` call. What is the shape of the `pareto_Y` you're passing into these calls?

**Reply by kks2107** [reply]
Hi @saitcakmak. Yes, the error occurs when calling hv.compute(). The pareto_y shape is 4.  
```
-------------------------------
BotorchTensorDimensionError               Traceback (most recent call last)
[<ipython-input-11-4e7d8831c8fc>](https://xn0uk1z8c4d-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20221118-060045-RC00_489427357#) in <module>
----> 1 hvs_qehvi, train_x_qehvi, train_obj_qehvi = mobo_execute(SEED, N_ITERATIONS, N_INITIAL)

1 frames
[<ipython-input-7-1c81d99955e6>](https://xn0uk1z8c4d-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20221118-060045-RC00_489427357#) in mobo_execute(seed, iterations, initial_samples)
     38 
     39 
---> 40     volume = hv.compute(pareto_y)
     41     hvs_qehvi.append(volume)
     42     save_xy(train_x_qehvi, train_obj_qehvi, hvs_qehvi)

[/usr/local/lib/python3.7/dist-packages/botorch/utils/multi_objective/hypervolume.py](https://xn0uk1z8c4d-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20221118-060045-RC00_489427357#) in compute(self, pareto_Y)
    144         if pareto_Y.shape[-1] != self._ref_point.shape[0]:
    145             raise BotorchTensorDimensionError(
--> 146                 "pareto_Y must have the same number of objectives as ref_point. "
    147                 f"Got {pareto_Y.shape[-1]}, expected {self._ref_point.shape[0]}."
    148             )

BotorchTensorDimensionError: pareto_Y must have the same number of objectives as ref_point. Got 4, expected 2.
```

If `num_objs` are the same size as `problem_dim`, then there is no error. But it should also work for different sizes.

**Reply by saitcakmak** [reply]
So, in this case your reference point has only 2 outcomes / objectives, but the `pareto_y` has 4 objectives. This is not about the mismatch between `num_objs` and `problem_dim` but the mismatch between `num_objs` (which is used to set the `ref_point`) and the actual number of objectives you have in `pareto_y`. If you correct that, it should work. 

Note that the performance of qNEHVI is sensitive to the choice of the reference point. It will try to discover the Pareto frontier that dominates the reference point. If you do not have a good enough prior on how to set the reference point, you can use the `infer_reference_point` heuristic with a set of observations.
`from botorch.utils.multi_objective.hypervolume import infer_reference_point`

**Reply by kks2107** [reply]
Thank you! I already thought that was the problem.
But it is possible that `num_objs = 2` and `problem_dim = 4`?

**Reply by saitcakmak** [reply]
> But it is possible that num_objs = 2 and problem_dim = 4?

Yes. The `problem_dim` here would be the dimension of the inputs `X` and `num_objs` is the dimension of the objectives `Y`. The two are independent since you can have a problem mapping an arbitrary D dimensional input space to M objectives.
